third_party/amd/lib/TritonAMDGPUTransforms/PointerCanonicalizer.cpp

```diff
@@ -0,0 +1,513 @@
+    Value nextPtr;
+
+    rewriter.setInsertionPoint(curOp);
+    if (auto splatOp = dyn_cast<triton::SplatOp>(curOp)) {
+      nextPtr = splatOp.getResult();
+      auto fatPtr = pointers[splatOp.getSrc()];
+      auto outType = splatOp.getResult().getType();
+      auto ptrShape = outType.getShape();
+      auto newOffsetType = RankedTensorType::get(
+          ptrShape, fatPtr.offset.getType(), outType.getEncoding());
+      Value offset = rewriter.create<triton::SplatOp>(curLoc, newOffsetType,
+                                                      fatPtr.offset);
+      // The shape of the fat pointer is contained within the offset. We don't
+      // need to keep the `splat` operation here.
+      opToDelete.insert(splatOp);
+      pointers[nextPtr] = FatPtr{splatOp.getSrc(), offset, fatPtr.canNarrow};
+    } else if (auto broadcastOp = dyn_cast<triton::BroadcastOp>(curOp)) {
+      nextPtr = broadcastOp.getResult();
+      auto fatPtr = pointers[broadcastOp.getSrc()];
+      auto outType =
+          dyn_cast<RankedTensorType>(broadcastOp.getResult().getType());
+      auto ptrShape = outType.getShape();
+      auto offsetType = dyn_cast<RankedTensorType>(fatPtr.offset.getType());
+      if (!offsetType)
+        return failure();
+
+      opToDelete.insert(broadcastOp);
+
+      auto newOffsetType = RankedTensorType::get(
+          ptrShape, offsetType.getElementType(), outType.getEncoding());
+      Value offset = rewriter.create<triton::BroadcastOp>(curLoc, newOffsetType,
+                                                          fatPtr.offset);
+      pointers[nextPtr] = FatPtr{fatPtr.basePtr, offset, fatPtr.canNarrow};
+
+    } else if (auto addPtr = dyn_cast<triton::AddPtrOp>(curOp)) {
+      nextPtr = addPtr.getResult();
+      auto fatPtr = pointers[addPtr.getPtr()];
+      Value fatPtrOffset = fatPtr.offset;
+      Value addPtrOffset = addPtr.getOffset();
+
+      Type fatPtrOffsetType = getElementType(fatPtrOffset);
+      Type addPtrOffsetType = getElementType(addPtrOffset);
+      Value newOffset = fatPtrOffset;
+      Value newPtr = fatPtr.basePtr;
+      bool canNarrow = fatPtr.canNarrow;
+      if (Value scalarOffset =
+              getScalarOffset(rewriter, curLoc, addPtrOffset)) {
+        // Scalar pointer update
+        newPtr = rewriter.create<triton::AddPtrOp>(curLoc, newPtr.getType(),
+                                                   newPtr, scalarOffset);
+      } else {
+        // If we the incoming offset is 32 bits, then we have to cast to 64
+        if (addPtrOffsetType.isInteger(32))
+          addPtrOffset =
+              extend32bitOffsetTo64Bits(rewriter, curLoc, addPtrOffset);
+
+        newOffset =
+            rewriter.create<arith::AddIOp>(curLoc, addPtrOffset, fatPtrOffset);
+
+        // If we do : %newOffset = %offset(32-bit) + %zero(64-bit) the user of
+        // %newOffset should always be able to narrow it down to 32 bits
+        bool isFatPtrZero = fatPtrOffset.getDefiningOp() &&
```

Comment:
Hmm, I'm not sure the logic is just this simple. I think we basically need a dedicated utility function to deduce whether we can narrow.. Cause clearly we will add more and more cases likely there.

Reply:
Unfortunately yes for now we cannot narrow in any other situation. If someone could tell us more then we can make the overflow logic more complex.